request_test.go

```diff
@@ -247,3 +249,108 @@ func TestOpendirHandleReuse(t *testing.T) {
 	rpkt = request.call(handlers, pkt, nil, 0)
 	assert.IsType(t, &sshFxpNamePacket{}, rpkt)
 }
+
+func Test_toLocalPath(t *testing.T) {
+	type args struct {
+		workDir string
+		p       string
+	}
+	tests := []struct {
+		name string
+		goos string
+		args args
+		want string
+	}{
+		{
+			name: "empty path",
+			goos: "linux",
+			args: args{p: ""},
+			want: "",
+		},
+		{
+			name: "relative path",
+			goos: "linux",
+			args: args{p: "file"},
+			want: "file",
+		},
+		{
+			name: "absolute path",
+			goos: "linux",
+			args: args{p: "/file"},
+			want: "/file",
+		},
+		{
+			name: "empty path",
+			goos: "linux",
+			args: args{workDir: "/home/user", p: ""},
+			want: "/home/user",
+		},
+		{
+			name: "relative path",
+			goos: "linux",
+			args: args{workDir: "/home/user", p: "file"},
+			want: "/home/user/file",
+		},
+		{
+			name: "relative path with .",
+			goos: "linux",
+			args: args{workDir: "/home/user", p: "."},
+			want: "/home/user",
+		},
+		{
+			name: "absolute path",
+			goos: "linux",
+			args: args{workDir: "/home/user", p: "/file"},
+			want: "/file",
+		},
+		{
+			name: "empty path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: ""},
+			want: "C:\\Users\\User",
+		},
+		{
+			name: "relative path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "file"},
+			want: "C:\\Users\\User\\file",
+		},
+		{
+			name: "relative path with .",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "."},
+			want: "C:\\Users\\User",
+		},
+		{
+			name: "absolute path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "C:\\file"},
+			want: "C:\\file",
+		},
+		{
+			name: "relative unix-like path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "./file"},
+			want: "C:\\Users\\User\\file",
+		},
+		{
+			name: "absolute unix-like path",
+			goos: "windows",
+			args: args{workDir: "C:\\Users\\User", p: "/C:/file"},
+			want: "C:\\file",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("%s %s %s", tt.goos, tt.args.workDir, tt.name), func(t *testing.T) {
```

Comment:
I don’t know why we wouldn’t just pre-compose all information necessary for the test into `tt.name` rather than sprinting it every time.

Reply:
It would duplicate information and introduce potential for mismatch between test name and what is actually being tested.

Reply:
The name of a test should concisely define the entire contract of the test. This is the programmer expressing the _intent_ of the test. If the test fails to meet this intent, then there is a bug in the tests.

Is there a bug in these tests?
```
tests := []test{
  {
    name: "test if",
    input: 2,
    expect: 2,
  }
}

for _, tt := range tests {
  tt.Run(fmt.Sprintf("%s %d = %d", tt.name, tt.input, tt.expect),  func(t *testing.T) {
    require.Equal(t, tt.input, tt.expect)
  })
}
```

What about here:
```
tests := []test{
  {
    name: "test if 3 = 2",
    input: 2,
    expect: 2,
  }
}

for _, tt := range tests {
  tt.Run(fmt.Sprintf("%s %d = %d", tt.name, tt.input, tt.expect),  func(t *testing.T) {
    require.Equal(t, tt.input, tt.expect)
  })
}
```

Better yet, what about here:
```
tests := []test{
  {
    name: "test if 3 = 2",
    input: 3,
    expect: 2,
  }
  {
    name: "test if 2 = 2",
    input: 2,
    expect: 2,
  }
}

for _, tt := range tests {
  tt.Run(fmt.Sprintf("%s %d = %d", tt.name, tt.input, tt.expect),  func(t *testing.T) {
    t.Parallel()
    require.Equal(t, tt.input, tt.expect)
  })
}
```

Reply:
I think I understand the point you you're making, but I don't see how this example is applicable. There'd be no difference in making a static name like `test if 2 = 3` vs using `fmt.Sprintf`. The main bug is that `tt` wasn't shadowed in the loop when using parallel tests and thus `require` getting called with the wrong arguments. (I mean the require arguments are also flipped, but I don't think that's the point you were making since all the information is already there for fixing it.)

In any case, I'll update the names :+1:.

Reply:
Yeah, I was trying to think of a proper example, thus the extensive edits of the comment. Brain wasn’t really working the best at the time after 10 hours of heavy debugging.

The require arguments being flipped is a bonus example of why I recommend against using the assert/require packages in general. It’s all too easy to flip arguments. But if I’m writing out the `if got != want { t.Errorf("got %d but wanted %d", want, got) }` then it’s way easier to see that the arguments are reversed.